lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,667 @@
+	ExceededNumberOfInputsOrOutputs,
+	InvalidTransactionState,
+	TransactionTooLarge,
+}
+
+//                   Interactive Transaction Construction negotiation
+//                           from the perspective of a holder
+//
+//                               AcceptingChanges
+//                        ┌──────────────────────────────┐
+//                        │                              │
+//                        │           ┌────────────────┐ │
+//                        │           │(sent/received) │ │
+//                        │           │tx_add_input    │ │
+//                        │           │tx_add_output   │ │
+//                        │           │tx_remove_input │ │
+//                        │           │tx_remove_output│ │
+//                        │           └───┐       ┌────┘ │
+//                        │               │       ▼      │
+//            ────────────┼──────────►┌───┴───────────┐  │        received_tx_complete                   ┌─────────────────────┐
+//    accept_channel2     │           │               ├──┼───────────────────┐          sent_tx_complete │                     │
+// or splice_ack          │     ┌─────┤  Negotiating  │  │                   ▼          ┌───────────────►│ NegotiationComplete │◄──┐
+// or tx_ack_rbf          │     │     │               │  │          ┌─────────────────┐ │                │                     │   │
+//    (sent or received)  │     │ ┌──►└───────────────┘  │          │                 │ │                └─────────────────────┘   │
+//                              │ │                      │          │ TheirTxComplete ├─┘                                          │
+//             sent_tx_complete │ │ received_tx_add_*    │          │                 │                   ┌────────────────────┐   │
+//                              │ │ received_tx_remove_* │          └─────────────────┘                   │                    │   │
+//                        │     │ │                      │                                            ┌──►│ NegotiationAborted │   │
+//                        │     │ └───┬───────────────┐  │        (sent/received)_tx_abort            │   │                    │   │
+//                        │     │     │               │  ├────────────────────────────────────────────┘   └────────────────────┘   │
+//                        │     └────►│ OurTxComplete │  │                                                                         │
+//                        │           │               ├──┼──┐                                                                      │
+//                        │           └───────────────┘  │  └──────────────────────────────────────────────────────────────────────┘
+//                        │                              │                         received_tx_complete
+//                        │                              │
+//                        └──────────────────────────────┘
+//
```

Comment:
Okay thanks for the ASCII, it’s helpful :) from looking on the interactive-tx spec, it sounds you have few messages missing such as `tx_init_rbf` and `tx_signatures` ? I think it’s good to have such two phases (fee-bumping, transaction valid and well-defined) in the state machines as they drive things like when to forgot a UTXO if we have more splice-in demand than UTXO available in the future.

Reply:
@dunxen I think we had discussed a bit on this initially and thought that the state machine would only handle up to mutual `tx_complete`s.  thoughts on this?

Reply:
Hmm I'd need to think how that might look here. Let's bring it up in the meet today. I think we can look at wrapping this state machine as I still believe it must only relate to a single negotiation and would be replaced with a fresh object if a new negotiation starts. The wrapper could keep track of previous tx_signatures, txs, bumping, as needed.

Reply:
A channel would hold onto that wrapper which contains a single active negotiation, but also has the other context mentioned above.